062-support-tiered-storage.md

```diff
@@ -0,0 +1,149 @@
+#  Support tiered storage natively in Strimzi
+
+## Current situation
+
+Kafka introduced tiered storage feature since 3.6. The feature is in early stage and it’s nice to support tiered storage configuration natively in Strimzi.
+
+## Motivation
+
+Supporting tiered storage in Strimzi would allow Strimzi users to enable tiered storage natively with Strimzi config. By allowing to config custom configuration of Kafka's `RemoteStorageManager` interface, Strimzi users would be able to easily adopt the tiered storage feature on Strimzi easily. More details about RemoteStorageManager can be found in [KIP405](https://cwiki.apache.org/confluence/display/KAFKA/KIP-405%3A+Kafka+Tiered+Storage).
+
+For simplicity, the proposal doesn't add support to customize `RemoteLogMetadataManager`. If tiered storage is enabled, the default implementation class `TopicBasedRemoteLogMetadataManager` in Kafka will be used.
+
+This proposal focuses on tiered storage support with a custom plugin brought in by the user. Other proposals in the future might add some built-in support for selected storage implementations such as S3. It should also be noted that, it’s on the end-user to ensure that their tiered storage setup works with Strimzi; the operator itself is strictly responsible for pushing down the necessary config to the broker pods.
+
+## Proposal
+
+Add a `spec.kafka.tieredStorage` field in the `Kafka` CR, which would support configuration for all tiered storage related configuration.
+
+The `tieredStorage` config contains configuration for some selected configuration of tiered storage feature. Most importantly, the configuration for `RemoteStorageManager` is needed.
+
+```
+kafka:
+  tieredStorage:
+    type: customRemoteStorageManager
+    className: <ClassName>
+    classPath: <ClassPath>
+    config:
+      # A map with String keys and String values. Key fields will be automatically prefixed with `rsm.config.` and appended to Kafka broker config.
+      <key>: <value>
```

Comment:
This allows configuring the RSM, but it doesn't look like it provides an obvious place to add configuration for the RLMM should we later want to support other implementations of that. 

* You could just shoehorn in some additional properties alongside the `type` etc:
    ```yaml
    tieredStorage:
      type: customRemoteStorageManager
      className: <ClassName>
      classPath: <ClassPath>
      config: ...
      rlmmClassName: <ClassName>
      rlmmConfig: ...
    ```
    but this looks quite ugly, imho.
* You could add an extra layer of nesting, so that tieredStorage` has a property for RSM (which is an object with `type`, `className` etc) and a separate property for the RLMM.
* You could just have a `remoteLogMetadataManager` property alongside the `tieredStorage` one. 

My understanding is that these two interfaces are independent, so I think I'd favour the last one, but with a slight twist: Rather than `tieredStorage` it should be `remoteStorageManager`, since that's all it's configuring: 

```yaml
remoteStorageManager: # I've just renamed this from tieredStorage
  type: customRemoteStorageManager
  className: <ClassName>
  classPath: <ClassPath>
  config: ...
remoteLogMetadataManager: # not part of this Proposal, but possibly in a future Proposal
  className: <ClassName>
  config: ...
```

Reply:
As stated in the proposal. We decided not to support custom RLMM here. That is why it is called `customRemoteStorageManager`.

Reply:
I mean, we can discuss it again if needed of course. But I think this is the assumption under which it was approved.

Reply:
My point is that "configuring tiered storage" boils down to configuring two things: the RSM and the RLMM. It's fine that this proposal only tackles one of those and assumes the default RLMM implementation. But it looks like the proposed CR API won't accommodate customising the RLMM in the future without it looking awkward. It would be counter-intuitive for the RLMM to be configured outside the `tieredStorage` object. And it also looks a bit weird, imho, if you try to add properties within the `tieredStorage` object (my first bullet point). 

To be clear, all I'm suggesting is that `tieredStorage` might be better named `remoteStorageManager`, since that's what it's configuring.

Reply:
I think the APIs for custom things will be always ugly. The way I imagine it, we start with:

```yaml
tieredStorage:
  type: customRemoteStorageManager
  # ...
```

But in the future we can easily add new type for both custom RSM and RLLM:

```yaml
tieredStorage:
  type: custom
  # ...
```

or:

```yaml
tieredStorage:
  type: s3
  # ...
```

----

What you are proposing would not end up with

```yaml
remoteStorageManager:
  # ...
```

and

```yaml
remoteLogMetadataManager:
  # ...
```

But we will likely add also add some options for S3 tiered storage and HDFS tiered storage etc. which will look completely different. I do not think this is a better API than what is already in the proposal.

----

Maybe they can be somehow combined? We do today something like this:

```yaml
  tieredStorage:
    type: custom
    remoteStorageManager:
      className: <ClassName>
      classPath: <ClassPath>
      config:
        # A map with String keys and String values. Key fields will be automatically prefixed with `rsm.config.` and appended to Kafka broker config.
        <key>: <value>
```

And in the future we can extend it to:

```yaml
  tieredStorage:
    type: custom
    remoteStorageManager:
      className: <ClassName>
      classPath: <ClassPath>
      config:
        # A map with String keys and String values. Key fields will be automatically prefixed with `rsm.config.` and appended to Kafka broker config.
        <key>: <value>
    remoteLogMetadataManager:
      # If ^^^ is not configured, use the default RLMM
```

Would that work better? Hard to say. As far as I understood, there is no custom RLMM today and it is not clear what requirements it might have for configuration. But it would be a simple change and if we say that the `remoteLogMetadataManager` might be added int he future it will be easy to adjust the proposal.

Reply:
I like the nested `remoteStorageManager`, it's not too different than what we have proposed now and it makes the proposal more extensible in the future. I updated the proposal, PTAL

Reply:
Does that work for you @tombentley?

Reply:
I don't understand :disappointed::

```yaml
  tieredStorage:                         <1>
    type: custom                         <2>
    remoteStorageManager:
      className: <ClassName>
      classPath: <ClassPath>
      config:
        # A map with String keys and String values. Key fields will be automatically prefixed with `rsm.config.` and appended to Kafka broker config.
        <key>: <value>
    remoteLogMetadataManager:
      # If ^^^ is not configured, use the default RLMM
```

1. Why do we need `tieredStorage`? Why can't the `remoteStorageManager` and `remoteLogMetadataManager` be direct children of the `spec.kafka`?
2. What is the purpose of the `type: custom`? Do we really need it to support (in the future) RSMs built into the Strimzi images? We could just allow `classPath` to be omitted in that future, but make it required for now.

Reply:
Re 1) That is what we are configuring, or?
Re 2) I think it is highly likely that we will want some better configuration for the most popular _types_ of tiered storage, yes. Not sure what it has to do with classpath eing required or optional.

Reply:
I prefer to keep `tieredStorage` as an umbrella wrapper for all related configs. It's clean and easy to manage and evolve the API object in the future.

Reply:
I also agree, having `tieredStorage` as an unmbrella makes really clear what the user is configuring. When people talk about this feature they talk about "tiered storage". Then it comes to details to be configured like RSM and RLMM.